benches/end_to_end.rs

```diff
@@ -1,385 +1,472 @@
+            .request_body
+            .expect("request_chunks means request_body");
+        rt.spawn(async move {
+            for _ in 0..chunk_cnt {
+                tx.send_data(chunk.into()).await.expect("send_data");
+            }
+        });
+        body
+    } else {
+        opts.request_body
+            .map(Body::from)
+            .unwrap_or_else(Body::empty)
+    };
+    let mut req = Request::new(body);
+    *req.method_mut() = opts.request_method.clone();
+    *req.uri_mut() = url.clone();
+    req
+}
+
+//
+// Prepares a client (request_sender) for sending requests to the given addr
+//
+fn prepare_client(
+    rt: &tokio::runtime::Runtime,
+    opts: &Opts,
+    addr: &SocketAddr,
+) -> SendRequest<crate::Body> {
+    let mut builder = Builder::new();
+    builder
+        .http2_only(opts.http2)
+        .http2_initial_stream_window_size(opts.http2_stream_window)
+        .http2_initial_connection_window_size(opts.http2_conn_window)
+        .http2_adaptive_window(opts.http2_adaptive_window);
+
+    let (request_sender, connection) = rt.block_on(async move {
+        let target_stream = TcpStream::connect(addr).await.unwrap();
+
+        builder.handshake(target_stream).await.unwrap()
+    });
+    rt.spawn(async move {
+        if let Err(e) = connection.await {
+            panic!("Error in connection: {}", e);
+        }
+    });
+    request_sender
+}
+
+//
+// Spawns a server in the background
+//
+fn spawn_server(rt: &tokio::runtime::Runtime, opts: &Opts) -> SocketAddr {
+    let addr: SocketAddr = "127.0.0.1:0".parse().unwrap();
+    let tcp_listener = rt.block_on(async move { TcpListener::bind(addr).await.unwrap() });
+    let local_addr = tcp_listener.local_addr().unwrap();
+    let mut http = Http::new();
+    //configure http based on opts
+    http.http2_only(opts.http2)
+        .http2_initial_stream_window_size(opts.http2_stream_window)
+        .http2_initial_connection_window_size(opts.http2_conn_window)
+        .http2_adaptive_window(opts.http2_adaptive_window);
+
+    let (http2, http2_stream_window, http2_conn_window, http2_adaptive_window) = (
+        opts.http2,
+        opts.http2_stream_window,
+        opts.http2_conn_window,
+        opts.http2_adaptive_window,
+    );
+    rt.spawn(async move {
+        loop {
+            let (tcp_stream, addr) = tcp_listener.accept().await.unwrap();
+            eprintln!("New incoming connection: {:?}", addr);
+            tokio::task::spawn(
+                Http::new()
+                    .http2_only(http2)
+                    .http2_initial_stream_window_size(http2_stream_window)
+                    .http2_initial_connection_window_size(http2_conn_window)
+                    .http2_adaptive_window(http2_adaptive_window)
+                    .serve_connection(tcp_stream, service_fn(handle_request)),
+            );
+        }
+    });
+    local_addr
+}
+
+async fn handle_request(_req: Request<Body>) -> Result<Response<Body>, Infallible> {
```

Comment:
In this function you return the request body, but you should be returning the `response_body` from the `Opts` structs.